llvm/lib/Target/Xtensa/XtensaISelLowering.cpp

```diff
@@ -697,6 +718,126 @@ const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
     return "XtensaISD::PCREL_WRAPPER";
   case XtensaISD::RET:
     return "XtensaISD::RET";
+  case XtensaISD::SELECT_CC:
+    return "XtensaISD::SELECT_CC";
   }
   return nullptr;
 }
+
+//===----------------------------------------------------------------------===//
+// Custom insertion
+//===----------------------------------------------------------------------===//
+
+static int GetBranchKind(int Cond, bool &BrInv) {
+  switch (Cond) {
+  case ISD::SETEQ:
+    return Xtensa::BEQ;
+  case ISD::SETNE:
+    return Xtensa::BNE;
+  case ISD::SETLT:
+    return Xtensa::BLT;
+  case ISD::SETLE:
+    BrInv = true;
+    return Xtensa::BGE;
+  case ISD::SETGT:
+    BrInv = true;
+    return Xtensa::BLT;
+  case ISD::SETGE:
+    return Xtensa::BGE;
+  case ISD::SETULT:
+    return Xtensa::BLTU;
+  case ISD::SETULE:
+    BrInv = true;
```

Comment:
Illegal condition codes should be handled by `setCondCodeAction(CC, Expand)`.

Reply:
I'm sorry, you mean to add setCondCodeAction() for codes which are not supported for i32 type natively by Xtensa? Currently all basic integer condition codes are supported, only codes like SETOLE which are used with FP types are not handled. Maybe I missed something.

Reply:
If I'm reading `XtensaInstrInfo.td` correctly, there are no instructions corresponding to ULE and UGT condition codes. These codes should be legalized by `setCondCodeAction(..., Expand)` and removed from this switch.

Reply:
I added patterns for brcc with  ULE and UGT codes:
[ brcc operation](https://github.com/llvm/llvm-project/blob/ac634c8b08adef1da8114821fbb7ad8a70e994a2/llvm/lib/Target/Xtensa/XtensaInstrInfo.td#L419)
And the same approach is used in lowering SELECTCC. 
So, your suggestion is to remove such ULE/UGT handling in code and use setCondCodeAction(CC, Expand) instead?

Reply:
> So, your suggestion is to remove such ULE/UGT handling in code and use setCondCodeAction(CC, Expand) instead?

Yes. That would allow you to remove those patterns and BrInv.

Reply:
It's interesting that if I use setCondCodeAction(CC, Expand) code instead of patterns then llc generates a bit large code in some cases, for example I compiled following code with "O3" option:
`
define i32 @select_ugt(i32 %a, ptr %b) nounwind {
  %val1 = load i32, ptr %b
  %tst1 = icmp ugt i32 %a, %val1
  %val2 = select i1 %tst1, i32 %a, i32 %val1
  ret i32 %val2
}

Old llc version:

select_ugt:
	l32i	a8, a3, 0
	bltu	a8, a2, .LBB1_2
	or	a2, a8, a8
.LBB1_2:
	ret

New llc version:

select_ugt:
	or	a8, a2, a2
	l32i	a2, a3, 0
	bgeu	a2, a8, .LBB1_2
	or	a2, a8, a8
.LBB1_2:
	ret
`

Should we strive to always rely on general approaches to lower Operations(like use setCondCodeAction), or in some cases (like that) we can use target specific implementation?

Reply:
I implemented expand action for GT/UGT/LE/ULE code conditions and remove patterns.